### ISSUE TITLE
Retry request on ETIMEDOUT

### ISSUE THREAD

**Issue by kaganjd** [netlify/js-client#166]
**- Do you want to request a *feature* or report a *bug*?**
feature? enhancement?

**- What is the current behavior?**
We have an enterprise customer who's hitting errors like this (redacted for privacy) a few times every day, impacting 5-10% of their daily builds:
```
FetchError: request to https://api.netlify.com/api/v1/sites/xxxxxxx-xxxxxxx failed, reason:
    connect ETIMEDOUT 3x.xxx.xx.xxx:443
    Code: ETIMEDOUT
```
@kitop investigated and found that the error was coming from node-fetch. He says:
> Seems like that's happening because one of the requests from the client fails to connect to the API servers when polling for the deploy being ready. Some related issue: node-fetch/node-fetch#817
If I turn on my wifi and try to run a command from CLI it shows a similar error

**Request**
Can we retry when there's a timeout like this? It looks like we currently retry when hitting a rate limit but not when there's a network issue: https://github.com/netlify/js-client/blob/a6476c93632af20080ec8a502522f81efd96c53c/src/methods/index.js#L69-L88

**Additional context**
- https://github.com/netlify/bitballoon/issues/7559#issuecomment-693587240 (internal)

**Comment by ehmicky** [comment]
:+1: 

We should use `got`, which has [built-in support](https://github.com/sindresorhus/got#retry) for automatic retries.

**Comment by mheffner** [comment]
@erquhart Do you think Pod B could prioritize this work? A customer has been hitting it recently (https://github.com/netlify/bitballoon/issues/7520#issuecomment-704424483) failing their deploys.

**Comment by erezrokah** [comment]
Other than adding edge handlers deploy to the CLI (which should not be related), the other deploy related CLI change is https://github.com/netlify/cli/pull/1272, which might have the impact of deploying more files.

Not sure how that might be related to polling to when the deploy is ready, which should be retried regardless. 

Also related https://github.com/netlify/js-client/issues/157

**Comment by erquhart** [comment]
@erezrokah @ehmicky let's hit this first thing in the morning.

**Comment by ehmicky** [comment]
The deploy is failing here:

https://github.com/netlify/js-client/blob/00b3e939ef133a15db0e9ba8e4e873d748fd442b/src/deploy/util.js#L73-L106

This is part of this library's `api.deploy()` method. This happens when polling `GET /sites/{site_id}/deploys/{deploy_id}` once per minute until the response returns `{ "state": "ready" }`. This polling happens right after uploading files and printing the log line `Waiting for deploy to go live...`. 

This polling times out after 20 minutes ([see configurable option `deployTimeout`](https://github.com/netlify/js-client/blob/00b3e939ef133a15db0e9ba8e4e873d748fd442b/src/deploy/index.js#L21)). However, I do not think this 20 minutes timeout is what's causing the `ETIMEDOUT` above. Instead, it would create a `TimeoutError` (see code [here](https://github.com/sindresorhus/p-wait-for/blob/d4aca6d4f1f0935eb622382a82373d16a1321fdc/index.js#L35) and [there](https://github.com/sindresorhus/p-timeout/blob/dececa8aa3d377d92d0abd64ea46d0f7efcd0688/index.js#L5)), which is different from `ETIMEDOUT`. [`ETIMEDOUT` is thrown](https://nodejs.org/dist/latest-v14.x/docs/api/errors.html#errors_common_system_errors) from the Node.js core `http` module instead and indicates a timeout at the TCP or HTTP level:

> `ETIMEDOUT` (Operation timed out): A connect or send request failed because the connected party did not properly respond after a period of time. Usually encountered by `http` or `net`. Often a sign that a `socket.end()` was not properly called.

This is probably not a bug with `node-fetch` either. The GitHub issue in the `node-fetch` repository linked above has been answered by one of the maintainers, and it does not look like a bug on their part. In the lines of code at https://github.com/node-fetch/node-fetch/blob/eaff0094c4dfdd5b78711a8c4f1b61e33d282072/src/index.js#L93-L96, `node-fetch` is only forwarding the error from Node.js core `http` module.

My most likely guess is that when performing `GET /sites/{site_id}/deploys/{deploy_id}`, there is a server timeout, either for the TCP connection to start, or for the HTTP headers and/or body to be received. 

Node.js recently changed the HTTP timeout logic in `14.11.0` on the [15th of September](https://github.com/nodejs/node/blob/master/doc/changelogs/CHANGELOG_V14.md#14.11.0). This might be unrelated though, I am just mentioning in case we starting experiencing issues on that day.

Retrying on `ETIMEDOUT` might patch this problem, but it looks like the core problem might be different: either something wrong with the uploading logic of this library `api.deploy()` method, or with our API timing out on `GET /sites/{site_id}/deploys/{deploy_id}`.

**Comment by erezrokah** [comment]
Cloud this also happen on an unstable network connection? If so retrying makes regardless of the root cause.

**Comment by kschlarman** [comment]
My assumption is that if this were an issue with our endpoint flaking out (`GET /sites/{site_id}/deploys/{deploy_id}`), we'd have many more reports of this issue. So my hunch is that it might be an unstable network connection.

I think it's pretty reasonable to expect that our customers may have flaky network connections. For this reason, many client libraries I've worked with will retry timeouts or other connection errors. So I view this as an awesome new feature that will improve the robustness of our CLI deploys.

**Comment by ehmicky** [comment]
Good point @kschlarman.
We're going to add some retry logic to `ETIMEDOUT` on any HTTP request from this library.
I'm currently figuring out exactly which part of the HTTP request/response is timing out with this specific error message, so we know what's the best place to put this retry logic.

**Comment by ehmicky** [comment]
Fixed at #171.

**Comment by ehmicky** [comment]
This has been released in `4.8.0`.
